In [2]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Stream Summer 2023

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files

Pandas has a number of useful file reading tools. You can see them enumerated by typing **"pd.re"** and pressing `tab`. We'll be using **read_csv** today. Note that these file reading functions do all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [3]:
#Answer Here
a=pd.read_csv('sets/elections.csv')
a

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

We can use the **head command** to show only a few rows of a dataframe.

# heading
## heading2

In [4]:
# Answer Here
a.head(2)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   

           %  
0  57.210122  
1  42.789878

There is also a **tail command**.

In [191]:
#Answer Here
a.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.

In [192]:
#Answer Here
a['Year']

0      1824
1      1824
2      1828
3      1828
4      1832
       ... 
177    2016
178    2020
179    2020
180    2020
181    2020
Name: Year, Length: 182, dtype: int64

Alternately, we could have used the **set_index** commmand on the dataframe.

In [193]:
#Answer Here
a.set_index('%')

Year          Candidate                  Party  Popular vote Result
%                                                                             
57.210122  1824     Andrew Jackson  Democratic-Republican        151271   loss
42.789878  1824  John Quincy Adams  Democratic-Republican        113142    win
56.203927  1828     Andrew Jackson             Democratic        642806    win
43.796073  1828  John Quincy Adams    National Republican        500897   loss
54.574789  1832     Andrew Jackson             Democratic        702735    win
...         ...                ...                    ...           ...    ...
1.073699   2016         Jill Stein                  Green       1457226   loss
51.311515  2020       Joseph Biden             Democratic      81268924    win
46.858542  2020       Donald Trump             Republican      74216154   loss
1.177979   2020       Jo Jorgensen            Libertarian       1865724   loss
0.255731   2020     Howard Hawkins                  Green        405035   loss

[182 rows x 5 columns]

In [194]:
a.reset_index()

index  Year          Candidate                  Party  Popular vote  \
0        0  1824     Andrew Jackson  Democratic-Republican        151271   
1        1  1824  John Quincy Adams  Democratic-Republican        113142   
2        2  1828     Andrew Jackson             Democratic        642806   
3        3  1828  John Quincy Adams    National Republican        500897   
4        4  1832     Andrew Jackson             Democratic        702735   
..     ...   ...                ...                    ...           ...   
177    177  2016         Jill Stein                  Green       1457226   
178    178  2020       Joseph Biden             Democratic      81268924   
179    179  2020       Donald Trump             Republican      74216154   
180    180  2020       Jo Jorgensen            Libertarian       1865724   
181    181  2020     Howard Hawkins                  Green        405035   

    Result          %  
0     loss  57.210122  
1      win  42.789878  
2      win  56.203927  
3     loss  43.796073  
4      win  54.574789  
..     ...        ...  
177   loss   1.073699  
178    win  51.311515  
179   loss  46.858542  
180   loss   1.177979  
181   loss   0.255731  

[182 rows x 7 columns]

# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [195]:
#Answer Here
b=pd.read_csv('sets/duplicate_columns.csv')
b

name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana

## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [196]:
# Answer Here
print(b.loc[[0]])
print("\n",b.iloc[[0],[2]])
print("\n",b.loc[[0],["flavor"]])
print("\n",b.loc[0:1,["name.1"]])
print("\n",b.loc[[0,2],"name":"name.1"])
print("\n",b.loc[:,"name":"name.1"])

   name name.1   flavor
0  john  smith  vanilla

     flavor
0  vanilla

     flavor
0  vanilla

   name.1
0  smith
1   shan

     name    name.1
0   john     smith
2  fulan  alfulani

     name    name.1
0   john     smith
1  zhang      shan
2  fulan  alfulani
3   hong   gildong


The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [197]:
# Answer Here
b.loc[0,"name.1"]

'smith'

A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [198]:
# Answer Here
series = a["Candidate"]
series

0         Andrew Jackson
1      John Quincy Adams
2         Andrew Jackson
3      John Quincy Adams
4         Andrew Jackson
             ...        
177           Jill Stein
178         Joseph Biden
179         Donald Trump
180         Jo Jorgensen
181       Howard Hawkins
Name: Candidate, Length: 182, dtype: object

In [199]:
# Answer Here
series.to_frame()

Candidate
0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
4       Andrew Jackson
..                 ...
177         Jill Stein
178       Joseph Biden
179       Donald Trump
180       Jo Jorgensen
181     Howard Hawkins

[182 rows x 1 columns]

### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [200]:
# Answer Here
a.loc[[1,2,3]]

Year          Candidate                  Party  Popular vote Result  \
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   

           %  
1  42.789878  
2  56.203927  
3  43.796073

If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [201]:
#elections[0] #this does not work, try uncommenting this to see it fail in action, woo
a[0]

KeyError: 0

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [202]:
# Answer Here
c=pd.DataFrame([[101,"Usman"],[102,"Saif"],[103,"Ahmad"]],columns=["Roll No","Name"])
c

Roll No   Name
0      101  Usman
1      102   Saif
2      103  Ahmad

Creating DataFrames using **Dictionary**.

In [203]:
# Answer Here
dic=pd.DataFrame({"Roll No":[101,102,103],"Name":["Usman","Saif","Ahmad"]})
dic

Roll No   Name
0      101  Usman
1      102   Saif
2      103  Ahmad

## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [204]:
print(b)
print('\n')
print(b[[True,False,False,True]])

    name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana


   name   name.1   flavor
0  john    smith  vanilla
3  hong  gildong   banana


One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [205]:
#Answer Here
x = a['Result']=='win'
x

0      False
1       True
2       True
3      False
4       True
       ...  
177    False
178     True
179    False
180    False
181    False
Name: Result, Length: 182, dtype: bool

Compare the 'Party' column to the String 'Democratic' and Show results

In [206]:
#Answer Here
x = a['Party']=='Democratic'
x

0      False
1      False
2       True
3      False
4       True
       ...  
177    False
178     True
179    False
180    False
181    False
Name: Party, Length: 182, dtype: bool

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [207]:
#Answer Here
isWin=(a['Year']>1980) & (a['Result'] == "win")
a[isWin]

Year          Candidate       Party  Popular vote Result          %
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
162  2008       Barack Obama  Democratic      69498516    win  53.023510
168  2012       Barack Obama  Democratic      65915795    win  51.258484
173  2016       Donald Trump  Republican      62984828    win  46.407862
178  2020       Joseph Biden  Democratic      81268924    win  51.311515

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [208]:
#Answer Here
isWin=(a['Year']<2000) & (a['Year']>1980) & (a['Result']=='win')
a[isWin]

Year          Candidate       Party  Popular vote Result          %
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938

Show all 'Loss' results of Independent party

In [209]:
# Answer Here
a[a['Party']== 'Independent']

Year         Candidate        Party  Popular vote Result          %
121  1976   Eugene McCarthy  Independent        740460   loss   0.911649
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [210]:
# Answer Here
a[(a['Result']=='win')&(a['%']<50)]

Year          Candidate                  Party  Popular vote Result  \
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
16   1848     Zachary Taylor                   Whig       1360235    win   
20   1856     James Buchanan             Democratic       1835140    win   
23   1860    Abraham Lincoln             Republican       1855993    win   
33   1876   Rutherford Hayes             Republican       4034142    win   
36   1880     James Garfield             Republican       4453337    win   
39   1884   Grover Cleveland             Democratic       4914482    win   
43   1888  Benjamin Harrison             Republican       5443633    win   
47   1892   Grover Cleveland             Democratic       5553898    win   
70   1912     Woodrow Wilson             Democratic       6296284    win   
74   1916     Woodrow Wilson             Democratic       9126868    win   
100  1948       Harry Truman             Democratic      24179347    win   
117  1968      Richard Nixon             Republican      31783783    win   
140  1992       Bill Clinton             Democratic      44909806    win   
144  1996       Bill Clinton             Democratic      47400125    win   
152  2000     George W. Bush             Republican      50456002    win   
173  2016       Donald Trump             Republican      62984828    win   

             %  
1    42.789878  
16   47.309296  
20   45.306080  
23   39.699408  
33   48.471624  
36   48.369234  
39   48.884933  
43   47.858041  
47   46.121393  
70   41.933422  
74   49.367987  
100  49.601536  
117  43.565246  
140  43.118485  
144  49.296938  
152  47.974666  
173  46.407862

Show all 'win' results between 1980 to 2000

In [6]:
# Answer Here
a[(a['Year']<2000) & (a['Year']>1980) & (a["Result"]=='win')]

Year          Candidate       Party  Popular vote Result          %
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938

## Loc and iLoc

Show 5 enteries from start

In [212]:
# Answer Here
a.loc[0:4]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [213]:
#Answer Here
a.loc[[0,1,2,3,4,5],['Candidate','Party', 'Year']]

Candidate                  Party  Year
0     Andrew Jackson  Democratic-Republican  1824
1  John Quincy Adams  Democratic-Republican  1824
2     Andrew Jackson             Democratic  1828
3  John Quincy Adams    National Republican  1828
4     Andrew Jackson             Democratic  1832
5         Henry Clay    National Republican  1832

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [214]:
# Answer Here
a.loc[5:10,"Year":"Party"]

Year               Candidate                Party
5   1832              Henry Clay  National Republican
6   1832            William Wirt         Anti-Masonic
7   1836       Hugh Lawson White                 Whig
8   1836        Martin Van Buren           Democratic
9   1836  William Henry Harrison                 Whig
10  1840        Martin Van Buren           Democratic

If we provide only a **single label** for the column argument, we get back a **Series**.

In [215]:
# Answer Here
a["Party"]

0      Democratic-Republican
1      Democratic-Republican
2                 Democratic
3        National Republican
4                 Democratic
               ...          
177                    Green
178               Democratic
179               Republican
180              Libertarian
181                    Green
Name: Party, Length: 182, dtype: object

If we want a data frame instead and don't want to use to_frame, we can provide a **list** containing the column name.

In [145]:
# Answer Here
a[['Result']]

Result
0     loss
1      win
2      win
3     loss
4      win
..     ...
177   loss
178    win
179   loss
180   loss
181   loss

[182 rows x 1 columns]

If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [8]:
# Answer Here
a.loc[0]

Year                             1824
Candidate              Andrew Jackson
Party           Democratic-Republican
Popular vote                   151271
Result                           loss
%                           57.210122
Name: 0, dtype: object

In [ ]:
# Answer Here

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [151]:
# Answer Here
a.loc[[0,1]]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   

           %  
0  57.210122  
1  42.789878

Specify Rows and Columns as List to retrive specific enteries

In [160]:
# Answer Here
row=[1,2,3]
column=["Year",'Result']
a.loc[row,column]

Year Result
1  1824    win
2  1828    win
3  1828   loss

Boolean Series are also boolean arrays, so we can use the Boolean Array Selection from earlier using loc as well.

In [ ]:
# Answer Here

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [162]:
# Answer Here
mottos = pd.read_csv('sets/mottos.csv')
mottos

State                                              Motto  \
0          Alabama                      Audemus jura nostra defendere   
1           Alaska                                North to the future   
2          Arizona                                         Ditat Deus   
3         Arkansas                                     Regnat populus   
4       California                                    Eureka (Εὕρηκα)   
5         Colorado                                    Nil sine numine   
6      Connecticut                            Qui transtulit sustinet   
7         Delaware                           Liberty and Independence   
8          Florida                                    In God We Trust   
9          Georgia                        Wisdom, Justice, Moderation   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
11           Idaho                                      Esto perpetua   
12        Illinois                  State sovereignty, national union   
13         Indiana                          The Crossroads of America   
14            Iowa  Our liberties we prize and our rights we will ...   
15          Kansas                                Ad astra per aspera   
16        Kentucky                               Deo gratiam habeamus   
17       Louisiana                         Union, justice, confidence   
18           Maine                                             Dirigo   
19        Maryland                       Fatti maschii, parole femine   
20   Massachusetts          Ense petit placidam sub libertate quietem   
21        Michigan          Si quaeris peninsulam amoenam circumspice   
22       Minnesota                                   L'étoile du Nord   
23     Mississippi                                   Virtute et armis   
24        Missouri                      Salus populi suprema lex esto   
25         Montana                                        Oro y plata   
26        Nebraska                            Equality before the law   
27          Nevada                                All For Our Country   
28   New Hampshire                                   Live Free or Die   
29      New Jersey                             Liberty and prosperity   
30      New Mexico                                      Crescit eundo   
31        New York                                          Excelsior   
32  North Carolina                                   Esse quam videri   
33    North Dakota                    \nSerit ut alteri saeclo prosit   
34            Ohio                  With God, all things are possible   
35        Oklahoma                                 Labor omnia vincit   
36          Oregon                                Alis volat propriis   
37    Pennsylvania                  Virtue, liberty, and independence   
38    Rhode Island                                               Hope   
39  South Carolina          Dum spiro spero \nAnimis opibusque parati   
40    South Dakota                          Under God the people rule   
41       Tennessee                           Agriculture and Commerce   
42           Texas                                         Friendship   
43            Utah                                           Industry   
44         Vermont                       Stella quarta decima fulgeat   
45        Virginia                                Sic semper tyrannis   
46      Washington                                      Al-ki or Alki   
47   West Virginia                              Montani semper liberi   
48       Wisconsin                                            Forward   
49         Wyoming                                       Equal Rights   

                                          Translation        Language  \
0                          We dare defend our rights!           Latin   
1                                                   —         English   
2                                        God enriches           Latin 

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [163]:
#Answer Here
mottos.iloc[0:3, 0:3]

State                          Motto                 Translation
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!
1   Alaska            North to the future                           —
2  Arizona                     Ditat Deus                God enriches

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [5]:
mottos = pd.read_csv("/content/drive/MyDrive/mottos.csv")

In [ ]:
# Answer Here

Size of DataFrame

In [164]:
# Answer Here
mottos.size

250

The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [165]:
# Answer Here
mottos.shape

(50, 5)

Use describe function and extract the meaningful information from DataFrame

In [166]:
# Answer Here
mottos.describe

<bound method NDFrame.describe of              State                                              Motto  \
0          Alabama                      Audemus jura nostra defendere   
1           Alaska                                North to the future   
2          Arizona                                         Ditat Deus   
3         Arkansas                                     Regnat populus   
4       California                                    Eureka (Εὕρηκα)   
5         Colorado                                    Nil sine numine   
6      Connecticut                            Qui transtulit sustinet   
7         Delaware                           Liberty and Independence   
8          Florida                                    In God We Trust   
9          Georgia                        Wisdom, Justice, Moderation   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
11           Idaho                                      Esto perpetua   
12        Illinoi

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [170]:
# Answer Here
mottos.index

RangeIndex(start=0, stop=50, step=1)

In [171]:
mottos.head(2)

State                          Motto                 Translation  \
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1   Alaska            North to the future                           —   

  Language Date Adopted  
0    Latin         1923  
1  English         1967

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [172]:
# Answer Here
mottos.columns

Index(['State', 'Motto', 'Translation', 'Language', 'Date Adopted'], dtype='object')

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [173]:
# Answer Here
mottos.sort_values

<bound method DataFrame.sort_values of              State                                              Motto  \
0          Alabama                      Audemus jura nostra defendere   
1           Alaska                                North to the future   
2          Arizona                                         Ditat Deus   
3         Arkansas                                     Regnat populus   
4       California                                    Eureka (Εὕρηκα)   
5         Colorado                                    Nil sine numine   
6      Connecticut                            Qui transtulit sustinet   
7         Delaware                           Liberty and Independence   
8          Florida                                    In God We Trust   
9          Georgia                        Wisdom, Justice, Moderation   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
11           Idaho                                      Esto perpetua   
12        Il

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [174]:
a.sort_values('%', ascending=False)

Year           Candidate         Party  Popular vote Result          %
114  1964      Lyndon Johnson    Democratic      43127041    win  61.344703
91   1936  Franklin Roosevelt    Democratic      27752648    win  60.978107
120  1972       Richard Nixon    Republican      47168710    win  60.907806
79   1920      Warren Harding    Republican      16144093    win  60.574501
133  1984       Ronald Reagan    Republican      54455472    win  59.023326
..    ...                 ...           ...           ...    ...        ...
165  2008    Cynthia McKinney         Green        161797   loss   0.123442
148  1996        John Hagelin   Natural Law        113670   loss   0.118219
160  2004    Michael Peroutka  Constitution        143630   loss   0.117542
141  1992            Bo Gritz      Populist        106152   loss   0.101918
156  2004          David Cobb         Green        119859   loss   0.098088

[182 rows x 6 columns]

We can also use `sort_values` on Series objects.

In [176]:
mottos['Language'].sort_values(ascending=False).head(50)

25           Spanish
18             Latin
47             Latin
45             Latin
44             Latin
39             Latin
36             Latin
35             Latin
33             Latin
32             Latin
31             Latin
30             Latin
24             Latin
23             Latin
21             Latin
20             Latin
0              Latin
11             Latin
15             Latin
2              Latin
3              Latin
6              Latin
16             Latin
5              Latin
19           Italian
10          Hawaiian
4              Greek
22            French
17           English
38           English
43           English
40           English
41           English
42           English
48           English
37           English
8            English
7            English
34           English
9            English
29           English
28           English
27           English
26           English
1            English
12           English
13           English
14           

For Series, the `value_counts` method is often quite handy.

In [185]:
mottos['Language'].value_counts()

Latin             23
English           21
Greek              1
Hawaiian           1
Italian            1
French             1
Spanish            1
Chinook Jargon     1
Name: Language, dtype: int64

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [186]:
mottos['Language'].unique()

array(['Latin', 'English', 'Greek', 'Hawaiian', 'Italian', 'French',
       'Spanish', 'Chinook Jargon'], dtype=object)

In [187]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(5)

5

# Thank you!